hypervisor/hw/pci.c

```diff
@@ -351,6 +351,11 @@ static void pci_read_cap(struct pci_pdev *pdev)
 				pdev->msix.table_offset = table_info & ~PCIM_MSIX_BIR_MASK;
 				pdev->msix.table_count = (msgctrl & PCIM_MSIXCTRL_TABLE_SIZE) + 1U;
 
+				if (pdev->msix.table_count > CONFIG_MAX_MSIX_TABLE_NUM) {
+					pr_fatal("Table count for PCI device with BDF 0x%04x is higher than the configured maximum (%d > %d)!",
+							pdev->bdf, pdev->msix.table_count, CONFIG_MAX_MSIX_TABLE_NUM);
+				}
+
```

Comment:
From FuSa p.o.v, this is deadcode and must be removed.  We should make it as pre-assumption and let FuSa test case make sure it.

We are defining standard error processing method with FuSa ..

Reply:
I am sorry but could you provide a link to the meaning of FuSa? I can not seem to recall the term.

I agree that this code is not contributing to any features. However, it helps identify a case where the compile-time configuration is not matching the system the hypervisor is intended to run on and will lead to pagefaults. Also this code only runs once for each physical PCI device which max 96 times with default configuration.

Is there a way this could comply to FuSa? As in having it terminate the hypervisor or something along those lines?

Sorry again for my ignorance of not knowing FuSa.

Reply:
@wuxyintel  can you suggest if we should use @pre here or debug only code or normal code? We may publish the error condition processing  rule to outside too.

Fusa means functional safety certification.

Reply:
Hi Eddie, in this case, "pdev->msix.table_count" is read from physical PCI configuration space, as we known, in the current code, physical PCI is configured by bootloader, the hypervisor read them through IO port. So physical PCI configuration information is external resource provided by bootloader. Suggest to check configuration error during platform initialization, if it detects some errors then panic, and define some constraints in the safety manual. But in this case, we maybe need to consider more other error cases and do checking, it is hard to do complete checking for physical PCI configuration information. For example, in this patch, this checking can't guarantee that "pdev->msix.table_count" is right. BTW, can we consider to statically define all physical PCI configuration of the platform and verify them manually?

Reply:
Currently in pci.c, during startup it will scan the physical pci bus ans store the results (bar base, bar size, capacity structure, etc) into global variables, then this info is used by other vpci modules.
In the future, we can statically defining all the pci config rather than scanning it dynamically, or  during startup, compare the statically defined pci confi with these dynamically retrieved results and throw an error if an mismatch is found, etc

Reply:
@DunderRoffe After reconsider more about this case from safety view, physical PCI configuration is provided by Board Support Package or firmware,  it is hard to do complete checking for physical PCI configuration information by hypervisor, so it is better to define application constraints on the hypervisor safety manual and Board Support Package or firmware should guarantee physical PCI configuration meet related application constraints, then the hypervisor can trust this configuration. 
So in this case, for debug purpose, we can use ASSERT to check whether MSI-X table_count
is larger than CONFIG_MAX_MSIX_TABLE_NUM or not.

Reply:
So you mean change to
```c
assert(pdev->msix.table_count > CONFIG_MAX_MSIX_TABLE_NUM);
```
instead of the if statement?

Reply:
@DunderRoffe  Please use assert(pdev->msix.table_count <= CONFIG_MAX_MSIX_TABLE_NUM), thanks!

Reply:
Updated per your request :)